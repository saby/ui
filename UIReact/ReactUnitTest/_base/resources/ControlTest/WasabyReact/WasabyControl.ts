import { Control, TemplateFunction } from 'UI/Base';

// @ts-ignore
import template = require('wml!ReactUnitTest/_base/resources/ControlTest/WasabyReact/WasabyControl');

export default class WasabyControl extends Control {
    readonly _template: TemplateFunction = template;
}
