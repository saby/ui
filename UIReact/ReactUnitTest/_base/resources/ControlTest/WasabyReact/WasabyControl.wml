<div>
    {{ _options.caption }}
</div>
