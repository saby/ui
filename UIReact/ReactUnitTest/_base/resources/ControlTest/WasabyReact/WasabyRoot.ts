import { Control, TemplateFunction } from 'UI/Base';

// @ts-ignore
import template = require('wml!ReactUnitTest/_base/resources/ControlTest/WasabyReact/WasabyRoot');

export default class WasabyRoot extends Control {
    readonly _template: TemplateFunction = template;
}
