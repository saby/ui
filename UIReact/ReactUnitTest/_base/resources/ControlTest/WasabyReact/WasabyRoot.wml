<div>
    <ws:partial template="{{_options.control}}" />
</div>
