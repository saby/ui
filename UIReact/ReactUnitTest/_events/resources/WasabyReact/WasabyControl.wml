<div on:click="_onClick()" id="{{ _options._id }}">
    {{ _options.caption }} {{ value }}
</div>
