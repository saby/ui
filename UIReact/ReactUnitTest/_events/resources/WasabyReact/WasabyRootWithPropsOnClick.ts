import { Control, TemplateFunction } from 'UI/Base';

// @ts-ignore
import template = require('wml!ReactUnitTest/_events/resources/WasabyReact/WasabyRootWithPropsOnClick');

export default class WasabyRootWithPropsOnClick extends Control {
    readonly _template: TemplateFunction = template;
}
