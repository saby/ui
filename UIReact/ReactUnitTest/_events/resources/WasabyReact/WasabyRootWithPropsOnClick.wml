<div>
    <ReactUnitTest._events.resources.WasabyReact.WasabyChildWithPropsOnClick onClick="{{ true }}" />
    <ReactUnitTest._events.resources.WasabyReact.WasabyChildWithPropsOnClick onClick="{{ 123 }}" />
    <ReactUnitTest._events.resources.WasabyReact.WasabyChildWithPropsOnClick onClick="{{ 'test' }}" />
</div>
