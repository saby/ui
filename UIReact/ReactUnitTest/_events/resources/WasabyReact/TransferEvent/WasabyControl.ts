import { Control, TemplateFunction } from 'UI/Base';
import * as template from 'wml!ReactUnitTest/_events/resources/WasabyReact/TransferEvent/WasabyControl';

const eventQueue: string[] = [];

export default class WasabyControl extends Control {
    _template: TemplateFunction = template;
    _eventQueue: string[] = [];

    _clickHandler(): void {
        this.pushToQueue('click in wasaby');
    }

    _myClickHandler(): void {
        this.pushToQueue('my click in wasaby');
    }

    _clickHandlerInTemplate(): void {
        this.pushToQueue('click template in wasaby');
    }

    _mouseDownHandler(): void {
        this.pushToQueue('mousedown in wasaby');
    }

    pushToQueue(value: string): void {
        eventQueue.push(value);
    }

    _applyEventQueue(): void {
        this._eventQueue = eventQueue;
    }
}
