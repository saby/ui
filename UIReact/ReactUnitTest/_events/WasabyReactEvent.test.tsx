/**
 * @jest-environment jsdom
 */
import { createSandbox } from 'sinon';
import { act } from 'react-dom/test-utils';

import { Control } from 'UICore/Base';
import WasabyRoot from './resources/WasabyReact/WasabyRoot';
import WasabyRootWithPropsOnClick from './resources/WasabyReact/WasabyRootWithPropsOnClick';
import WasabyControl from './resources/WasabyReact/TransferEvent/WasabyControl';

const creator = Control.createControl;
const destroyer = Control.destroyControl;

describe.skip('WasabyReact Events', () => {
    let container;
    let sandbox;
    let clock;
    let consoleMock;
    let reactControlOnClickMock;
    let instance;

    beforeEach(() => {
        sandbox = createSandbox();
        clock = sandbox.useFakeTimers();
        consoleMock = jest.spyOn(console, 'error').mockImplementation();
        container = document.createElement('div');
        document.body.appendChild(container);
        reactControlOnClickMock = jest.fn();
    });

    afterEach(() => {
        clock.restore();
        sandbox.restore();
        consoleMock.mockRestore();
        destroyer(instance, container);
        container.remove();
        container = null;
        reactControlOnClickMock.mockClear();
    });

    function tick(duration: number): void {
        act(() => {
            clock.tick(duration);
        });
    }

    it('Проверяем события, если подписка добавлена через onClick на wasaby-контрол', () => {
        // region Setup
        act(() => {
            instance = creator(
                WasabyRoot,
                { clickFn: reactControlOnClickMock },
                container
            );
        });
        tick(0);
        // endregion
        const button = document.getElementById('wasabyControl');

        act(() => {
            button.dispatchEvent(
                new window.MouseEvent('click', { bubbles: true })
            );
        });
        tick(0);

        expect(reactControlOnClickMock).toHaveBeenCalledTimes(1);
    });

    it('Проверяем события, если подписка добавлена через onClick на wasaby-контрол (HOC)', () => {
        // region Setup
        act(() => {
            instance = creator(
                WasabyRoot,
                { clickFn: reactControlOnClickMock },
                container
            );
        });
        tick(0);
        // endregion
        const button = document.getElementById('wasabyControlHoc');
        act(() => {
            button.dispatchEvent(
                new window.MouseEvent('click', { bubbles: true })
            );
        });
        tick(0);

        expect(reactControlOnClickMock).toHaveBeenCalledTimes(1);
    });

    it('Проверяем события, если подписка добавлена через onMouseDown на wasaby-контрол (HOC)', () => {
        // region Setup
        act(() => {
            instance = creator(
                WasabyRoot,
                { clickFn: reactControlOnClickMock },
                container
            );
        });
        tick(0);
        // endregion
        const button = document.getElementById('wasabyControlHoc');
        act(() => {
            button.dispatchEvent(
                new window.MouseEvent('click', { bubbles: true })
            );
        });
        tick(0);

        expect(reactControlOnClickMock).toHaveBeenCalledTimes(1);
    });

    it('Проверяем события, если подписка добавлена через onClick на wasaby-контрол в корне которого partial с шаблоном', () => {
        // region Setup
        act(() => {
            instance = creator(
                WasabyRoot,
                { clickFn: reactControlOnClickMock },
                container
            );
        });
        tick(0);
        // endregion
        const button = document.getElementById('wasabyControlTemplate');
        act(() => {
            button.dispatchEvent(
                new window.MouseEvent('click', { bubbles: true })
            );
        });
        tick(0);

        expect(reactControlOnClickMock).toHaveBeenCalledTimes(1);
    });

    it('Проверяем события, если подписка добавлена через onMouseDown на wasaby-контрол в корне которого partial с шаблоном', () => {
        // region Setup
        act(() => {
            instance = creator(
                WasabyRoot,
                { clickFn: reactControlOnClickMock },
                container
            );
        });
        tick(0);
        // endregion
        const button = document.getElementById('wasabyControlTemplate');
        act(() => {
            button.dispatchEvent(
                new window.MouseEvent('click', { bubbles: true })
            );
        });
        tick(0);

        expect(reactControlOnClickMock).toHaveBeenCalledTimes(1);
    });

    it('Проверяем события, react-обработчик использует аргумент event', () => {
        // region Setup
        act(() => {
            instance = creator(
                WasabyRoot,
                { clickFn: reactControlOnClickMock },
                container
            );
        });
        tick(0);
        // endregion
        const button = document.getElementById('wasabyControlWithArgs');
        act(() => {
            button.dispatchEvent(
                new window.MouseEvent('click', { bubbles: true })
            );
        });
        tick(0);

        expect(reactControlOnClickMock).toHaveBeenCalledTimes(1);
    });

    it('Проверяем события, react-обработчик использует аргумент event', () => {
        // region Setup
        act(() => {
            instance = creator(
                WasabyRoot,
                { clickFn: reactControlOnClickMock },
                container
            );
        });
        tick(0);
        // endregion
        const button = document.getElementById('wasabyControlWithArgs2');
        act(() => {
            button.dispatchEvent(
                new window.MouseEvent('click', { bubbles: true })
            );
        });
        tick(0);

        expect(reactControlOnClickMock).toHaveBeenCalledTimes(1);
    });

    it('Проверяем события, если именование опции совпадает с именем реакт-события, но не является обработчиком', () => {
        // region Setup
        act(() => {
            instance = creator(WasabyRootWithPropsOnClick, {}, container);
        });
        tick(0);
        // endregion
        const button = document.getElementById('testButton');
        act(() => {
            button.dispatchEvent(
                new window.MouseEvent('click', { bubbles: true })
            );
        });
        tick(0);

        expect(consoleMock).not.toHaveBeenCalled();
        expect(container).toMatchSnapshot();
    });

    it('Проверяем события, порядок вызова обработчиков react-wasaby', () => {
        // region Setup
        act(() => {
            instance = creator(WasabyControl, {}, container);
        });
        tick(0);
        // endregion
        const reactButton = document.getElementById('reactElement');
        const checkButton = document.getElementById('checkButton');
        act(() => {
            reactButton.dispatchEvent(
                new window.MouseEvent('mousedown', { bubbles: true })
            );
        });
        tick(15);
        act(() => {
            reactButton.dispatchEvent(
                new window.MouseEvent('mouseup', { bubbles: true })
            );
        });
        tick(15);
        act(() => {
            checkButton.dispatchEvent(
                new window.MouseEvent('click', { bubbles: true })
            );
        });
        tick(15);

        expect(container).toMatchSnapshot();
    });
});
