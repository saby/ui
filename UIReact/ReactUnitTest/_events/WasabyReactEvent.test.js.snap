// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WasabyReact Events Проверяем события, если именование опции совпадает с именем реакт-события, но не является обработчиком 1`] = `
<div
  style="height: 100%; width: 100%; display: contents;"
>
  <div>
    <button
      id="testButton"
      tabindex="0"
    >
      true
    </button>
    <button
      id="testButton"
      tabindex="0"
    >
      123
    </button>
    <button
      id="testButton"
      tabindex="0"
    >
      test
    </button>
  </div>
</div>
`;

exports[`WasabyReact Events Проверяем события, порядок вызова обработчиков react-wasaby 1`] = `
<div
  style="height: 100%; width: 100%; display: contents;"
>
  <div>
    <div
      id="reactElement"
      tabindex="0"
    >
      react control
    </div>
    <button
      id="checkButton"
    >
      check
    </button>
    <div>
      <div>
        mousedown in wasaby
      </div>
      <div>
        click in react
      </div>
      <div>
        click template in wasaby
      </div>
      <div>
        click in wasaby
      </div>
    </div>
  </div>
</div>
`;
